### PR TITLE
terml rewrite

### PR DESCRIPTION
**Pull request by jml**
This reworks the branch to use terml, and to have a two-stage separation between parser and interpreting, and between mapping Python to edn ('decoding') and unparsing.
## notes for reviewer
1. `_extended._Decoder` and `_ast._Builder` have roughly the same shape and
   similar problems (`leafData` is weird, shouldn't even need to handle
   `'.tuple.'`, dispatch is inevitably on `tag.name`, `term` method is almost
   guaranteed useless).  Muck around a bit with making a better direct API for
   term traversal in terml.
2. Maybe leave `decode` policy-free, passing in `_DECODERS` and
   `DEFAULT_READERS` and leave the default policy shenanigans up to `loads`
3. In a later branch, I want to refactor `encode` to have a separate, generic
   Python object traversal thing.  I need to sit down and think about this one.
4. Perhaps create an `unknown_handler` for `encode`, and also base rules for
   encoding supported types like int, float, and so forth.
5. Doesn't support numbers as well as it could.  I think this can wait for a
   later branch.
6. Almost 100% of the extension is going to be turning a domain-specific object
   into a tagged value and back.  We need to make sure there's a good API for
   that.
7. `constantly` belongs in perfidy
8. Default encoding rules map tuples to lists and lists to vectors, basically
   because tuples use parens and lists use brackets.
## questions for reviewer
1. How do you feel about the module break up?
2. Can you come up with better names than `encode` and `decode`?
3. Should there be exposed `Keyword` and `Symbol` objects that are _not_
   terms, but rather representing actual keywords and symbols?  Would make
   some things (e.g. type-based dispatch in `encode`) easier.
   - Maybe give `Symbol` a `makeTag` method, so that `INST.makeTag("2013-...")` works?
4. What should `encode` do with `TaggedValue`s?
5. I haven't really thought much about the public interface. Is anything public that should be private? Private that should be public?

### CODE REVIEW

**Comment by jml** [comment]
Broken. Will try again later.

**Comment by jml** [comment]
Actually was working. `paredit` pebkac.

**Comment by jml** [comment]
Relies on https://github.com/python-parsley/parsley/pull/43 to work.

File: edn/_ast.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+def parse(string):
+    return edn(string).edn()
+
+
+def _dump_bool(obj):
+    if obj:
+        return 'true'
+    else:
+        return 'false'
```
Comment (jml) [inline edn/_ast.py]: Not used any more. Delete.

File: edn/_ast.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        'TaggedValue': TaggedValue,
+        'Map': Map,
+        'Nil': Nil,
+        'Set': Set,
+        'List': List,
+    },
+    name='edn')
+
+
+def parse(string):
```
Comment (jml) [inline edn/_ast.py:58]: Docstring.

File: edn/_ast.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
+from functools import partial
```
Comment (jml) [inline edn/_ast.py:19]: Module docstring needed.

Also, perhaps this module should be made publicly available as `edn.ast` (or `edn.syntax` or what-have-you) and then top-level `edn` can export the object-level functions.

File: edn/_ast.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        return getattr(self, '_dump_%s' % (tag.name,))
+
+    def leafData(self, data, span=None):
+        for base_type, dump_rule in self.PRIMITIVES:
+            if isinstance(data, base_type):
+                return lambda args: dump_rule(data)
+        raise ValueError("Cannot encode %r" % (data,))
+
+    def term(self, f, built_terms):
+        return f(built_terms)
```
Comment (jml) [inline edn/_ast.py]: If we change this to `f(*built_terms)`, it might simplify some of the other code above.

**Comment by jml** [comment]
All my comments addressed. Waiting on your review.

**Comment by dreid** [comment]
First I'll answer your questions:
1. Considering everything is private I'm not too concerned about the way the modules are broken up at this point, this layout looks as reasonable as any other.
2. `to_terms`/`from_terms`, I don't know that I feel strongly about it, I think the public API has to be `load`/`loads` and `dump`/`dumps` no matter what?  Should functions for getting to the intermediate state be public at all?
3. That seems reasonable.  You probably also need to do things like support comparisons for namespaceless symbols and keywords to strings.  I'm OK tackling that in separate pull requests though?
4. Assuming unknown tagged values, i'm inclined to say that the default should be to raise an exception.  The only other options are `repr(o)` some protocol for asking an object to encode itself `o.to_edn` or actually supporting the pickle protocol.  It seems obvious that `dump*` should support a `default` keyword like in `json`.  An alternative is to not let unknown `TaggedValues` escape from `load*`.  This would maintain symmetry between the APIs and let us not deal with what the default encoding is.
5. I'm cool with everything but `loads` and `dumps` (and Symbol and Keyword) being private.

**Comment by dreid** [comment]
Comments on some of your notes:
2. I like the idea of `encode` and `decode` not being very opinionated about what to do with extension types.  This sounds a lot like how `json` is laid out with it's `object_hook` kwarg.
4. `json.dump*` has a `default` kwarg which takes an un-encodable object and returns an encodable object.
8. The only alternative I see to this is to actually implement proper singly-linked list for lists and PersistentVectors for vectors.  If that's the road one wants to take it should be taken sooner rather than later.

**Comment by jml** [comment]
Questions:
1. Agreed.
2. Let's leave intermediate functions private for now. We _might_ want streaming variants of `load*`/`dump*`. At present, `loads` especially only understands a single edn element, but the format is streaming. This is a key difference from json, and maybe worth _only_ having streaming APIs to emphasise this. Streaming is a separate pull request.
3. Definitely separate pull request.
4. Agreed. Default should be to raise an exception and the handler of defaults should be parametrizable.
5. Agreed.

Notes:
2. Agreed.
4. Yes. See Q4 answer.
8. I want to take that road soon -- very soon. There's a limit to how much unmerged code I want to have at any one time though.

File: edn/_ast.py (hunk 3 above)
Comment (dreid) [inline edn/_ast.py:19]: Yes to module docstring, eventually maybe it should be public, but I wouldn't worry about it very much yet.

Comment (dreid) [inline edn/_ast.py:19]: The module docstring should also do it's best to document the terml representations produced by this thing.

**Comment by jml** [comment]
(for my own benefit)

### actions from questions & notes
- [x] rename `encode` to `to_terms`
- [x] rename `decode` to `from_terms`
- [x] change top-level exports so only `load*`, `dump*`, `Symbol` and `Keyword` are public
- [x] add `default` handler to `decode` / `from_terms`

File: edn/_ast.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        return 'false'
+
+    def _dump_Nil(self):
+        return 'nil'
+
+    def _dump_Character(self, obj):
+        return '\\' + obj
+
+    def _dump_Keyword(self, obj):
+        return ':' + obj
```
Comment (dreid) [inline edn/_ast.py:92]: Does this account for obj being a `Symbol`?

File: edn/_extended.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    build = getattr(term, 'build', None)
+    if build:
+        return build(builder)
+    return builder.leafData(term)(term)
+
+
+def loads(string, readers=frozendict(), default=None):
+    """Interpret an edn string.
+
+    :param term: A parsed edn term, probably got from `edn.parse`.
```
Comment (dreid) [inline edn/_extended.py]: This does it's own parsing, it doesn't take a term.

File: edn/_extended.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    :param readers: A map from tag symbols to callables.  For '#foo bar'
+        whatever callable the Symbol('foo') key is mapped to will be
+        called with 'bar'.  There are default readers for #inst and #uuid,
+        which can be overridden here.
+    :return: Whatever the string is interpreted as.
```
Comment (dreid) [inline edn/_extended.py]: This is… not an awesome description of what the return value is.

File: edn/_extended.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    :return: Whatever term gets decoded to.
+    """
+    builder = _Decoder(_DECODERS, DEFAULT_READERS.merge(readers), default)
+    build = getattr(term, 'build', None)
+    if build:
+        return build(builder)
+    return builder.leafData(term)(term)
+
+
+def loads(string, readers=frozendict(), default=None):
```
Comment (dreid) [inline edn/_extended.py:119]: default is undocumented.

**Comment by dreid** [comment]
Other than those rather simple inline notes I don't have a lot of input on this.  Everything seems to be the right shape  and the test coverage is good.  I didn't see anything obviously wrong.  After you address those 4 outstanding notes feel free to merge (I _think_ you can do that.)

File: edn/_ast.py (hunk 5 above)
Comment (jml) [inline edn/_ast.py:92]: Yes, it's magically recursive thanks to the magic of `Builder`

File: edn/_extended.py (hunk 6 above)
Comment (jml) [inline edn/_extended.py]: Fixed

File: edn/_extended.py (hunk 7 above)
Comment (jml) [inline edn/_extended.py]: Agreed. I've changed it, but I doubt it'll ever inspire awe.

**Pull request merged**